src/main/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperations.scala

```diff
@@ -156,64 +143,94 @@ object TezosDatabaseOperations {
-    } yield (o.fee, o.timestamp)
-    val op = dbHandle.run(action.distinct.sortBy(_._2.desc).take(numberOfFeesAveraged).result)
-    val results = Await.result(op, Duration.apply(awaitTimeInSeconds, SECONDS))
-    val resultNumbers = results.map(x => x._1.getOrElse("0").toDouble)
-    val m: Int = ceil(mean(resultNumbers)).toInt
-    val s: Int = ceil(stdev(resultNumbers)).toInt
-    results.length match {
-      case 0 => None
-      case _ =>
-        val timestamp = results.head._2
-        Some(AverageFees(max(m - s, 0), m, m + s, timestamp, kind))
+  def calculateAverageFeesIO(kind: String)(implicit ec: ExecutionContext): DBIO[Option[AverageFees]] = {
+    def computeAverage(ts: java.sql.Timestamp, fees: Seq[(Option[String], java.sql.Timestamp)]): AverageFees = {
+      val values = fees.map {
+        case (fee, _) => fee.map(_.toDouble).getOrElse(0.0)
+      }
+      val m: Int = ceil(mean(values)).toInt
+      val s: Int = ceil(stdev(values)).toInt
+      AverageFees(max(m - s, 0), m, m + s, ts, kind)
     }
-  }
 
+    val opQuery =
+      Tables.Operations
+        .filter(_.kind === kind)
+        .map(o => (o.fee, o.timestamp))
+        .distinct
+        .sortBy { case (_, ts) => ts.desc }
+        .take(numberOfFeesAveraged)
+        .result
+
+    opQuery.map { 
+      timestampedFees =>
+        timestampedFees.headOption.map {
+          case (_, latest) =>
+            computeAverage(latest, timestampedFees)
+        }
+    }
+  }
 
   /**
     * Delete all accounts in database not associated with block at maxLevel.
-    * @return No value, database update
+    * @return the number of rows removed
     */
-  def purgeOldAccounts(): Try[Int] = {
-    ApiOperations.fetchMaxBlockLevelForAccounts().flatMap{ blockLevel =>
-      Try {
-        val query = Tables.Accounts.filter(row => !(row.blockLevel === blockLevel))
-        val action = query.delete
-        val op = dbHandle.run(action)
-        Await.result(op, Duration.apply(awaitTimeInSeconds, SECONDS))
-      }
+  def purgeOldAccounts()(implicit ex: ExecutionContext): Future[Int] = {
+    val purged = dbHandle.run {
+      val maxLevel = SupportQueries.accountsMaxBlockLevel
+      Tables.Accounts.filter(_.blockLevel =!= maxLevel).delete
     }
+    purged.andThen {
+      case Success(howMany) => logger.info("{} accounts where purged from old block levels.", howMany)
+      case Failure(e) => logger.error("Could not purge old block-levels accounts", e)
+    }
+  }
+
+  /** convertions from domain objects to database row format */
+  object RowConversion {
+
+    private[TezosDatabaseOperations] def ofAverageFees(in: AverageFees) =
+      Tables.FeesRow(
+        low = in.low,
+        medium = in.medium,
+        high = in.high,
+        timestamp = in.timestamp,
+        kind = in.kind
+      )
+
   }
 
+  object SupportQueries {
+
+    /* use as max block level when none exists */
+    private[TezosDatabaseOperations] val defaultBlockLevel: BigDecimal = -1
```

Comment:
A lot of code we use relies on the original version of this query. Why does putting this in a separate object help? Would rewriting everything that relies on maxBlockLevel to use this SupportQueries object be fruitful?

Reply:
Please ignore this part, I later figured out that this will best fit into another class, and also that I got the semantic wrong, when using the `ifNull`, so I'll need to rewrite this specific piece of code.

Let's discuss  that after the change to see if it works better

Reply:
Got it.